ipatests/test_integration/test_epn.py

```diff
@@ -494,3 +497,126 @@ def test_EPN_delay_config(self, cleanupmail):
         self.master.put_file_contents('/etc/ipa/epn.conf', epn_conf)
         result = tasks.ipa_epn(self.master, raiseonerr=False)
         assert "smtp_delay cannot be less than zero" in result.stderr_text
+
+    def generate_user_batch(self):
+        """
+        Prepare large (10k) batch of users with various attributes, namely
+        without passwords, gradually expiring passwords, unicode names and
+        unicode e-mails
+        :return: None
+        """
+        USERS_IN_GROUP = 2000
+        users = deque()  # entries are stored as tuple (username, expiration)
+
+        # Users without passwords
+        for i in range(USERS_IN_GROUP):
+            username = "user_wo_pass_%d" % i
+            tasks.user_add(self.master, username)
+            users.append((username, -1))
+
+        # Users with passwords gradually expiring
+        for i in range(USERS_IN_GROUP):
+            username = "user_with_pass_%d" % i
+            # TODO: Compute
+            krbpasswordexpiration = datetime_to_generalized_time(
+                datetime.datetime.utcnow() + datetime.timedelta(days=7)
+            )
+            tasks.user_add(
+                self.master,
+                username,
+                password="Secret123",
+                extra_args=["--password-expiration",
+                            krbpasswordexpiration],
+            )
+            users.append((username, krbpasswordexpiration))
+
+        # Users with wide unicode in usernames and with passwords to expire
+        all_langs = hg.Languages.get_all()
+        users_per_lang = USERS_IN_GROUP // len(all_langs)
+        remainder = USERS_IN_GROUP % len(all_langs)
+        krbpasswordexpiration = datetime_to_generalized_time(
+            datetime.datetime.utcnow() + datetime.timedelta(days=7)
```

Comment:
Similar to above where this is outside the loop but you probably still only need to compute the time once since we're checking in full days. Plus depending on the TOD that this test executes it could have wonky results if some users were pre midnight and some were post.